src/main/client/batch_write.c

```diff
@@ -265,29 +271,37 @@ static PyObject *AerospikeClient_BatchWriteInvoke(AerospikeClient *self,
         long return_type = -1;
 
         as_operations *ops = NULL;
-        if (py_ops_size) {
+        if (batch_type == AS_BATCH_READ ||
+            batch_type == AS_BATCH_WRITE &&
+                (py_ops_size || (py_meta != NULL && py_meta != Py_None))) {
+
             ops = as_operations_new(py_ops_size);
             garb->ops_to_free = ops;
-        }
-
-        for (Py_ssize_t i = 0; i < py_ops_size; i++) {
 
-            PyObject *py_op = PyList_GetItem(py_ops_list, i);
-            if (py_op == NULL || !PyDict_Check(py_op)) {
-                as_error_update(
-                    err, AEROSPIKE_ERR_PARAM,
-                    "py_op is NULL or not a dict, %s must be a dict \
-                                 produced by an aerospike operation helper",
-                    FIELD_NAME_BATCH_OPS);
-                goto CLEANUP0;
+            if (py_meta) {
+                if (check_for_meta(py_meta, ops, err) != AEROSPIKE_OK) {
```

Comment:
Rename `check_for_meta()` to `check_and_set_meta()` since this function writes metadata to `ops`,